tests/frontier/precompiles/test_precompiles.py

```diff
@@ -0,0 +1,103 @@
+
+    Args:
+        fork (Fork): The fork instance containing precompiled contract information.
+
+    Yields:
+        Iterator[Tuple[str, bool]]: A tuple containing the address in hexadecimal format and a
+            boolean indicating whether the address is a supported precompile.
+
+    """
+    supported_precompiles = fork.precompiles()
+
+    num_unsupported = NUM_UNSUPPORTED_PRECOMPILES
+    for address in range(1, UPPER_BOUND + 1):
+        if address in supported_precompiles:
+            yield (hex(address), True)
+        elif num_unsupported > 0:
+            # Check unsupported precompiles up to NUM_UNSUPPORTED_PRECOMPILES
+            yield (hex(address), False)
+            num_unsupported -= 1
+
+
+@pytest.mark.valid_from("Berlin")
+@pytest.mark.parametrize_by_fork("address,precompile_exists", precompile_addresses)
+def test_precompiles(
+    state_test: StateTestFiller, address: str, precompile_exists: bool, pre: Alloc
+):
+    """
+    Tests the behavior of precompiled contracts in the Ethereum state test.
+
+    Args:
+        state_test (StateTestFiller): The state test filler object used to run the test.
+        address (str): The address of the precompiled contract to test.
+        precompile_exists (bool): A flag indicating whether the precompiled contract exists at the
+            given address.
+        pre (Alloc): The allocation object used to deploy the contract and set up the initial
+            state.
+
+    This test deploys a contract that performs two CALL operations to the specified address and a
+    fixed address (0x10000), measuring the gas used for each call. It then stores the difference
+    in gas usage in storage slot 0. The test verifies the expected storage value based on
+    whether the precompiled contract exists at the given address.
+
+    """
+    env = Environment()
+
+    account = pre.deploy_contract(
+        Op.MSTORE(0x00, Op.GAS)
+        + Op.CALL(address=address)
+        + Op.MSTORE(0x00, Op.SUB(Op.MLOAD(0x00), Op.GAS))
+        + Op.MSTORE(0x20, Op.GAS)
+        + Op.CALL(address=0x10000)
+        + Op.MSTORE(0x20, Op.SUB(Op.MLOAD(0x20), Op.GAS))
+        + Op.SSTORE(
+            0,
+            Op.LT(
+                Conditional(
+                    condition=Op.GT(Op.MLOAD(0x00), Op.MLOAD(0x20)),
+                    if_true=Op.SUB(Op.MLOAD(0x00), Op.MLOAD(0x20)),
+                    if_false=Op.SUB(Op.MLOAD(0x20), Op.MLOAD(0x00)),
+                ),
+                0x1A4,
+            ),
+        )
+        + Op.STOP,
+        storage={0: 0xDEADBEEF},
+    )
+
+    tx = Transaction(
+        to=account,
+        sender=pre.fund_eoa(),
+        gas_limit=1_000_000,
+        protected=True,
+    )
+
+    # A high gas cost will result from calling a precompile
+    # Expect 0x00 when a precompile exists at the address, 0x01 otherwise
+    post = {account: Account(storage={0: "0x00" if precompile_exists else "0x01"})}
```

Comment:
here I don't clearly understand why abs(empty_call_gas - precompile_call_gas) < 0x1a4 means this logic. 
if you can rewrite it more clean like 
`if precompile_exists then precompile_call_gas  is x` else y  if it is possible.

Reply:
I'm not clear on this, since the actual gas value varies by precompile call. I'm not sure what logic could be set outside of the contract code. Could you explain a bit more on this?